pkg/commands/exec_live_default.go

```diff
@@ -0,0 +1,121 @@
+// +build !windows
+
+package commands
+
+import (
+	"bufio"
+	"os"
+	"os/exec"
+	"strings"
+	"sync"
+	"unicode/utf8"
+
+	"github.com/kr/pty"
+	"github.com/mgutz/str"
+)
+
+// RunCommandWithOutputLiveWrapper runs a command and return every word that gets written in stdout
+// Output is a function that executes by every word that gets read by bufio
+// As return of output you need to give a string that will be written to stdin
+// NOTE: If the return data is empty it won't written anything to stdin
+// NOTE: You don't have to include a enter in the return data this function will do that for you
+func RunCommandWithOutputLiveWrapper(c *OSCommand, command string, output func(string) string) (errorMessage string, codeError error) {
+	cmdOutput := []string{}
+
+	splitCmd := str.ToArgv(command)
+	cmd := exec.Command(splitCmd[0], splitCmd[1:]...)
+
+	cmd.Env = os.Environ()
+	cmd.Env = append(cmd.Env, "LANG=en_US.utf8", "LC_ALL=en_US.UTF-8")
+
+	tty, err := pty.Start(cmd)
+
+	if err != nil {
+		return "", err
+	}
+
+	stopAsking := make(chan struct{})
+
+	var waitForBufio sync.WaitGroup
+	waitForBufio.Add(1)
+
+	defer func() {
+		_ = tty.Close()
+	}()
+
+	go func() {
+		scanner := bufio.NewScanner(tty)
+		scanner.Split(scanWordsWithNewLines)
+		for scanner.Scan() {
+			select {
+			case <-stopAsking:
+				// just do nothing
```

Comment:
just doing nothing here didn't look right to me, so I've done some testing and it looks like we can just remove this channel altogether. I believe that when the cmd is finished, the tty sends an EOF marker to the scanner, making it stop. That means that the scanner.Scan() loop finishes of its own accord, so we can then print out the output.

This patch worked fine for me, no hanging at all:
```
diff --git a/pkg/commands/exec_live_default.go b/pkg/commands/exec_live_default.go
index 5fb84c8..418baf4 100644
--- a/pkg/commands/exec_live_default.go
+++ b/pkg/commands/exec_live_default.go
@@ -34,38 +34,25 @@ func RunCommandWithOutputLiveWrapper(c *OSCommand, command string, output func(s
 		return "", err
 	}
 
-	stopAsking := make(chan struct{})
-
 	var waitForBufio sync.WaitGroup
 	waitForBufio.Add(1)
 
-	defer func() {
-		_ = tty.Close()
-	}()
-
 	go func() {
 		scanner := bufio.NewScanner(tty)
 		scanner.Split(scanWordsWithNewLines)
 		for scanner.Scan() {
-			select {
-			case <-stopAsking:
-				// just do nothing
-			default:
-				toOutput := strings.Trim(scanner.Text(), " ")
-				cmdOutput = append(cmdOutput, toOutput)
-				toWrite := output(toOutput)
-				if len(toWrite) > 0 {
-					_, _ = tty.Write([]byte(toWrite + "\n"))
-				}
+			toOutput := strings.Trim(scanner.Text(), " ")
+			cmdOutput = append(cmdOutput, toOutput)
+			toWrite := output(toOutput)
+			if len(toWrite) > 0 {
+				_, _ = tty.Write([]byte(toWrite + "\n"))
 			}
 		}
 		waitForBufio.Done()
 	}()
 
 	err = cmd.Wait()
-	go func() {
-		stopAsking <- struct{}{}
-	}()
+	tty.Close()
 	if err != nil {
 		waitForBufio.Wait()
 		return strings.Join(cmdOutput, " "), err

```

Reply:
hmm doing some more testing with this patch it appears that pulling works but pushing doesn't

Reply:
testing with both mine and your changes the seem issue seems to arise:

If you don't enter anything for the username and then you hit enter, the credentials panel goes away leaving just the confirmation panel saying 'pushing...' but it never goes away. Looks like we need to handle the case where there is an error after the username returning an error

Reply:
alright I've managed to iron out the bugs I found. I've put up a new branch called 'password-testing' with my changes on it. I'd push directly to this branch but it looks like changes have been made since I pulled so I don't want to accidentally blow away any of your changes

Reply:
Nice, I'll look into that :+1:

Reply:
Just tested your code and it seems to work great.  
I'll add it to this branch, this seems to me like a much better approach to this problem.  
Thx for doing some research.